### PR TITLE
Add: vault selector when creating assistant

### PR DESCRIPTION
**Pull request by Fraggle**
## Description

- Add an intermediary screen to pick a vault when having more than 1.
- Fix: admins can only use vaults where they are members (this [commit](https://github.com/dust-tt/dust/pull/7127/commits/f6499549b08d61eb4c952f6a55344d50c7bd080c))

Updated to take into account the global limit of 1 vault + company.

(will do another PR with some design revisions from @Duncid )

## Risk

None

## Deploy Plan

Deploy `front`

### CODE REVIEW

File: front/components/assistant_builder/ActionsScreen.tsx [hunk 1]
```diff
@@ -122,12 +132,65 @@ export default function ActionsScreen({
...
 }) {
+  const { vaults } = useContext(AssistantBuilderContext);
+
   const configurableActions = builderState.actions.filter(
     (a) => !(CAPABILITIES_ACTION_CATEGORIES as string[]).includes(a.type)
   );
 
   const isLegacyConfig = isLegacyAssistantBuilderConfiguration(builderState);
 
+  const vaultIdToActions = useMemo(() => {
```
Comment (flvndvd) [inline front/components/assistant_builder/ActionsScreen.tsx:143]: IIRC, @philipperolet introduced a `getDataSourceViewIdsFromActions` method, can we use it to ensure there is only one place where we need to perform this operation?

File: front/components/assistant_builder/ActionsScreen.tsx [hunk 2]
```diff
@@ -578,7 +654,40 @@ function ActionConfigEditor({
...
+  vaultsUsedInActions,
+  instructions,
+  updateAction,
+  setEdited,
+  description,
+  onDescriptionChange,
+}: ActionConfigEditorProps) {
+  const { vaults } = useContext(AssistantBuilderContext);
+
+  // Only allow global + one regular across all actions (weird product decision that we will fix later)
```
Comment (flvndvd) [inline front/components/assistant_builder/ActionsScreen.tsx:671]:
```suggestion
  // Only allow global + one regular across all actions (product decision that we will revisit later)
```

File: front/components/assistant_builder/AssistantBuilderDataSourceModal.tsx [hunk 3]
```diff
@@ -77,9 +80,28 @@ export default function AssistantBuilderDataSourceModal({
       variant="side-md"
       title="Manage data sources selection"
     >
-      <div className="w-full pt-12">
+      <div className="flex w-full justify-end pl-12 pr-2 pt-4">
```
Comment (flvndvd) [inline front/components/assistant_builder/AssistantBuilderDataSourceModal.tsx:83]: This is orthogonal to the vault effort, right? Instead of simulating those clicks we usually push the highest parent(s) in the array of selected resources, can we follow this existing pattern?

File: front/components/assistant_builder/actions/ProcessAction.tsx [hunk 4]
```diff
@@ -254,34 +254,41 @@ function PropertiesFields({
...
-  | {
-      description?: undefined;
-      onDescriptionChange?: undefined;
-    }
-)) {
+  description: string;
+  onDescriptionChange: (description: string) => void;
+};
+
+export function ActionProcess({
```
Comment (flvndvd) [inline front/components/assistant_builder/actions/ProcessAction.tsx:272]: Not required right now but we will have to apply the same logic on Dust Apps.

File: front/components/assistant_builder/server_side_props_helpers.ts [hunk 5]
```diff
@@ -43,15 +43,16 @@ import logger from "@app/logger/logger";
 
 export const getAccessibleSourcesAndApps = async (auth: Authenticator) => {
   const accessibleVaults = [
-    await VaultResource.fetchWorkspaceGlobalVault(auth),
-  ];
+    ...(await VaultResource.listWorkspaceVaults(auth, false)),
+  ].filter((vault) => !vault.isSystem());
```
Comment (flvndvd) [inline front/components/assistant_builder/server_side_props_helpers.ts:47]: Might be worth a comment.

Comment (flvndvd) [reply front/components/assistant_builder/server_side_props_helpers.ts:47]: Created https://github.com/dust-tt/dust/issues/7146 to improve (not urgent).

File: front/components/data_source_view/DataSourceViewSelector.tsx [hunk 6]
```diff
@@ -29,30 +36,147 @@ import {
...
+  const dataSourceViewsByVaultId = useMemo(
+    () => _.groupBy(dataSourceViews, (dsv) => dsv.vaultId),
+    [dataSourceViews]
+  );
+
+  const { vaults, isVaultsError, isVaultsLoading } = useVaults({
+    workspaceId: owner.sId,
+  });
+
+  const [selectedVault, setSelectedVault] = useState<string>(
```
Comment (flvndvd) [inline front/components/data_source_view/DataSourceViewSelector.tsx:75]: I'd assume that the default selected vault should be the global one.

File: front/components/data_source_view/DataSourceViewSelector.tsx [hunk 7]
```diff
@@ -29,30 +36,147 @@ import {
...
+    Object.keys(selectionConfigurations).length > 0
+      ? selectionConfigurations[Object.keys(selectionConfigurations)[0]]
+          .dataSourceView.vaultId
+      : ""
+  );
+
+  if (isVaultsLoading || isVaultsError || !vaults) {
```
Comment (flvndvd) [inline front/components/data_source_view/DataSourceViewSelector.tsx:82]: `nit`: `useVaults` should always return an array and not `null`

File: front/components/data_source_view/DataSourceViewSelector.tsx [hunk 8]
```diff
@@ -29,30 +36,147 @@ import {
...
+                      </>
+                    ),
+                    value: vaultId,
+                    disabled: disabled,
+                  },
+                ]}
+                value={selectedVault}
+                onChange={() => setSelectedVault(vaultId)}
+              />
+              {selectedVault == vaultId && (
```
Comment (flvndvd) [inline front/components/data_source_view/DataSourceViewSelector.tsx:149]:
```suggestion
              {selectedVault === vaultId && (
```

File: front/components/data_source_view/DataSourceViewSelector.tsx [hunk 9]
```diff
@@ -29,30 +36,147 @@ import {
...
+      ? selectionConfigurations[Object.keys(selectionConfigurations)[0]]
+          .dataSourceView.vaultId
+      : ""
+  );
+
+  if (isVaultsLoading || isVaultsError || !vaults) {
+    return <Spinner />;
+  }
+
+  if (Object.keys(dataSourceViewsByVaultId).length === 1) {
```
Comment (flvndvd) [inline front/components/data_source_view/DataSourceViewSelector.tsx:86]: Is this to serve the case where the workspace does not have any vault (outside of the global one)?

File: front/components/data_source_view/DataSourceViewSelector.tsx [hunk 10]
```diff
@@ -76,96 +200,116 @@ export function DataSourceViewsSelector({
...
+              }
+              setSelectionConfigurations={setSelectionConfigurations}
+              viewType={viewType}
+            />
+          ))}
+
+        {groupFolders && (
+          <Tree.Item
+            key="folders"
+            label="Files"
```
Comment (flvndvd) [inline front/components/data_source_view/DataSourceViewSelector.tsx:266]:
```suggestion
            label="Folders"
```

File: front/components/data_source_view/DataSourceViewSelector.tsx [hunk 11]
```diff
@@ -212,8 +356,28 @@ function DataSourceViewSelector({
       ? "partial"
       : "unchecked";
 
-  const onToggleSelectAll = useMemo(() => {
-    return (checked: boolean) => {
+  const keepOnlyOneVaultIfApplicable = useCallback(
```
Comment (flvndvd) [inline front/components/data_source_view/DataSourceViewSelector.tsx:359]: Not sure to understand what we are trying to do here.

File: front/lib/resources/vault_resource.ts [hunk 12]
```diff
@@ -156,11 +156,14 @@ export class VaultResource extends BaseResource<VaultModel> {
   }
 
   static async listWorkspaceVaults(
-    auth: Authenticator
+    auth: Authenticator,
+    adminsBypassACL: boolean = true
```
Comment (flvndvd) [inline front/lib/resources/vault_resource.ts:160]: Let's not do this even more with a default value set to `true`. See https://github.com/dust-tt/dust/issues/7146.

File: front/components/assistant_builder/ActionsScreen.tsx (hunk 1 above)
Comment (Fraggle) [reply front/components/assistant_builder/ActionsScreen.tsx:143]: Just saw it after merging.

File: front/components/assistant_builder/ActionsScreen.tsx (hunk 2 above)
Comment (Fraggle) [reply front/components/assistant_builder/ActionsScreen.tsx:671]: This it outdated.

File: front/components/assistant_builder/AssistantBuilderDataSourceModal.tsx (hunk 3 above)
Comment (Fraggle) [reply front/components/assistant_builder/AssistantBuilderDataSourceModal.tsx:83]: >This is orthogonal to the vault effort, right ?

It's in the figma.

I don't understand "usually push the highest parent(s) in the array of selected resources"

Can you point me to some code ?

File: front/components/data_source_view/DataSourceViewSelector.tsx (hunk 9 above)
Comment (Fraggle) [reply front/components/data_source_view/DataSourceViewSelector.tsx:87]: yes

File: front/components/data_source_view/DataSourceViewSelector.tsx (hunk 11 above)
Comment (Fraggle) [reply front/components/data_source_view/DataSourceViewSelector.tsx:360]: Added comments.

File: front/components/assistant_builder/ActionsScreen.tsx (hunk 1 above)
Comment (Fraggle) [reply front/components/assistant_builder/ActionsScreen.tsx:143]: Looked..

Not super practical for that use case (it takes form body, returns datasourceviewids).
I'll keep as it.

File: front/components/assistant_builder/ActionsScreen.tsx [hunk 13]
```diff
@@ -122,12 +132,65 @@ export default function ActionsScreen({
...
+        case "PROCESS":
+          Object.values(action.configuration.dataSourceConfigurations).forEach(
+            (config) => {
+              if (config.dataSourceView) {
+                vaultIdToActions[config.dataSourceView.vaultId] = (
+                  vaultIdToActions[config.dataSourceView.vaultId] ?? []
+                ).concat(action);
+              }
+            }
+          );
```
Comment (flvndvd) [inline front/components/assistant_builder/ActionsScreen.tsx:176]:
```suggestion
        case "PROCESS":
        case "RETRIEVAL_SEARCH":
        case "RETRIEVAL_EXHAUSTIVE":
          Object.values(action.configuration.dataSourceConfigurations).forEach(
            (config) => {
              vaultIdToActions[config.dataSourceView.vaultId] = (
                vaultIdToActions[config.dataSourceView.vaultId] ?? []
              ).concat(action);
            }
          );
          break;
```

File: front/components/assistant_builder/ActionsScreen.tsx [hunk 14]
```diff
@@ -578,7 +654,35 @@ function ActionConfigEditor({
...
+  setEdited,
+  description,
+  onDescriptionChange,
+}: ActionConfigEditorProps) {
+  const { vaults } = useContext(AssistantBuilderContext);
+
+  // Only allow one vault across all actions
+  const allowedVaults = useMemo(() => {
+    const usedVaultsInOtherActions = vaults.reduce((acc, v) => {
+      return (vaultsUsedInActions[v.sId] ?? []).some((a) => a !== action)
```
Comment (flvndvd) [inline front/components/assistant_builder/ActionsScreen.tsx:674]: Can we use the action name instead of the reference?

File: front/components/assistant_builder/ActionsScreen.tsx [hunk 15]
```diff
@@ -578,7 +654,35 @@ function ActionConfigEditor({
...
+        ? acc.concat(v)
+        : acc;
+    }, [] as VaultType[]);
+
+    return vaults.filter(
+      (vault) =>
+        usedVaultsInOtherActions.length == 0 ||
+        usedVaultsInOtherActions.find((v) => v.sId === vault.sId)
+    );
+  }, [action, vaults, vaultsUsedInActions]);
```
Comment (flvndvd) [inline front/components/assistant_builder/ActionsScreen.tsx:684]: Pretty hard to read IMO, may I suggest something like:
```suggestion
const allowedVaults = useMemo(() => {
  const isVaultUsedInOtherActions = (vault: VaultType) => {
    const actionsUsingVault = vaultsUsedInActions[vault.sId] ?? [];
    return actionsUsingVault.some((a) => a.name !== action.name);
  };

  const usedVaultsInOtherActions = vaults.filter(isVaultUsedInOtherActions);

  if (usedVaultsInOtherActions.length === 0) {
    return vaults;
  }

  return vaults.filter((vault) =>
    usedVaultsInOtherActions.some((v) => v.sId === vault.sId)
  );
}, [action, vaults, vaultsUsedInActions]);
```

File: front/components/assistant_builder/actions/ProcessAction.tsx [hunk 16]
```diff
@@ -254,34 +254,41 @@ function PropertiesFields({
...
-  updateAction,
-  setEdited,
-  description,
-  onDescriptionChange,
-}: {
+type ActionProcessProps = {
   owner: WorkspaceType;
   instructions: string | null;
   actionConfiguration: AssistantBuilderProcessConfiguration | null;
+  allowedVaults: VaultType[];
```
Comment (flvndvd) [inline front/components/assistant_builder/actions/ProcessAction.tsx:261]: Given that we need to spread vaults in many places now, I'm wondering if we don't want to leverage the existing assistant builder context to avoid adding all those new props. Context can have state management function so we could think of a function to get the allowed vaults based on the current action, WDYT?

File: front/components/data_source_view/DataSourceViewSelector.tsx (hunk 6 above)
Comment (flvndvd) [reply front/components/data_source_view/DataSourceViewSelector.tsx:76]: If I understand correctly, here we pick the vault used by the first action, is that right? Why not defaulting to the global one?

File: front/components/data_source_view/DataSourceViewSelector.tsx [hunk 17]
```diff
@@ -30,30 +37,147 @@ import {
...
+  const { vaults, isVaultsError, isVaultsLoading } = useVaults({
+    workspaceId: owner.sId,
+  });
+
+  const [selectedVault, setSelectedVault] = useState<string>(
+    Object.keys(selectionConfigurations).length > 0
+      ? selectionConfigurations[Object.keys(selectionConfigurations)[0]]
+          .dataSourceView.vaultId
+      : ""
+  );
```
Comment (flvndvd) [inline front/components/data_source_view/DataSourceViewSelector.tsx:81]:
```suggestion
const [selectedVault, setSelectedVault] = useState<string>(() => {
  const firstKey = Object.keys(selectionConfigurations)[0];
  return firstKey ? selectionConfigurations[firstKey].dataSourceView.vaultId : "";
});
```

File: front/components/data_source_view/DataSourceViewSelector.tsx (hunk 9 above)
Comment (flvndvd) [reply front/components/data_source_view/DataSourceViewSelector.tsx:87]: Let's add a small comment for the next one that wonders 😉

File: front/components/data_source_view/DataSourceViewSelector.tsx [hunk 18]
```diff
@@ -30,30 +37,147 @@ import {
...
+              <RadioButton
+                name={`Vault ${vaultId}`}
+                choices={[
+                  {
+                    label: (
+                      <>
+                        <Icon
+                          visual={getVaultIcon(vault)}
+                          size="md"
+                          className="s-flex-shrink-0 ml-3 inline-block align-middle text-brand"
```
Comment (flvndvd) [inline front/components/data_source_view/DataSourceViewSelector.tsx:121]: Tailwind classes prefixed with `s-` should only be used in Sparkle 🙃 
```suggestion
                          className="flex-shrink-0 ml-3 inline-block align-middle text-brand"
```

File: front/components/data_source_view/DataSourceViewSelector.tsx [hunk 19]
```diff
@@ -30,30 +37,147 @@ import {
...
+                        />
+                        <span
+                          className={classNames(
+                            "s-text-element-900",
```
Comment (flvndvd) [inline front/components/data_source_view/DataSourceViewSelector.tsx:125]:
```suggestion
                            "text-element-900",
```

File: front/components/data_source_view/DataSourceViewSelector.tsx (hunk 10 above)
Comment (flvndvd) [reply front/components/data_source_view/DataSourceViewSelector.tsx:267]: Bump :)

File: front/components/data_source_view/DataSourceViewSelector.tsx (hunk 11 above)
Comment (flvndvd) [reply front/components/data_source_view/DataSourceViewSelector.tsx:360]: Did you forget to push 🤔

File: front/components/assistant_builder/ActionsScreen.tsx (hunk 14 above)
Comment (Fraggle) [reply front/components/assistant_builder/ActionsScreen.tsx:674]: I wanted to but the name is not unique, therefor I can't make sure which one are the "others" later.

File: front/components/assistant_builder/actions/ProcessAction.tsx (hunk 16 above)
Comment (Fraggle) [reply front/components/assistant_builder/actions/ProcessAction.tsx:261]: I'm not a fan, it make the "logic" spread between the global and local context. I think it will make it harder to follow the code flow.

File: front/components/data_source_view/DataSourceViewSelector.tsx (hunk 17 above)
Comment (Fraggle) [reply front/components/data_source_view/DataSourceViewSelector.tsx:81]: I think you still need to check the length but I'll update to improve lisibility.

File: front/components/data_source_view/DataSourceViewSelector.tsx (hunk 9 above)
Comment (Fraggle) [reply front/components/data_source_view/DataSourceViewSelector.tsx:87]: I did (but forgot to push :p)

File: front/components/assistant_builder/ActionsScreen.tsx (hunk 15 above)
Comment (Fraggle) [reply front/components/assistant_builder/ActionsScreen.tsx:684]: Done (with action vs action.name for the reason explained above)

File: front/components/assistant_builder/ActionsScreen.tsx (hunk 14 above)
Comment (flvndvd) [reply front/components/assistant_builder/ActionsScreen.tsx]: I guess the name isn't unique in the UI but for pretty sure that when saving it must be unique.

File: front/components/data_source_view/DataSourceViewSelector.tsx [hunk 20]
```diff
@@ -30,30 +37,150 @@ import {
...
+  });
+
+  const [selectedVault, setSelectedVault] = useState<string>(() => {
+    const firstKey =
+      Object.keys(selectionConfigurations).length > 0
+        ? Object.keys(selectionConfigurations)[0]
+        : null;
+    return firstKey
+      ? selectionConfigurations[firstKey].dataSourceView.vaultId
+      : "";
```
Comment (flvndvd) [inline front/components/data_source_view/DataSourceViewSelector.tsx:83]:
```suggestion
    const firstKey = Object.keys(selectionConfigurations)[0] ?? null;
    return firstKey ? selectionConfigurations[firstKey]?.dataSourceView?.vaultId ?? "" : "";
```

File: front/components/data_source_view/DataSourceViewSelector.tsx (hunk 9 above)
Comment (flvndvd) [reply front/components/data_source_view/DataSourceViewSelector.tsx:90]: Still don't see it 😛

File: front/components/data_source_view/DataSourceViewSelector.tsx [hunk 21]
```diff
@@ -77,96 +204,116 @@ export function DataSourceViewsSelector({
...
+      <VaultSelector
+        {...{
+          owner,
+          dataSourceViews,
+          allowedVaults,
+          selectionConfigurations,
+          setSelectionConfigurations,
+          viewType,
+        }}
+      />
```
Comment (flvndvd) [inline front/components/data_source_view/DataSourceViewSelector.tsx:218]:
```suggestion
      <VaultSelector
          owner={owner}
          dataSourceViews={dataSourceViews}
          allowedVaults={allowedVaults}
          selectionConfigurations={selectionConfigurations}
          setSelectionConfigurations={setSelectionConfigurations}
          viewType={viewType}
      />
```

File: front/components/data_source_view/DataSourceViewSelector.tsx [hunk 22]
```diff
@@ -213,8 +360,32 @@ function DataSourceViewSelector({
       ? "partial"
       : "unchecked";
 
-  const onToggleSelectAll = useMemo(() => {
-    return (checked: boolean) => {
+  // If the user has multiples vaults, they can choose to only select one vault per tool
+  // It's forced in the UI because it's a radio button
+  // However, when they select something in another vault, the previous selections will not be removed automatically
+  // This function is used to remove the all the previous selections that do not match the passed selection's vault
```
Comment (flvndvd) [inline front/components/data_source_view/DataSourceViewSelector.tsx:366]:
```suggestion
  // This function is used to remove all the previous selections that do not match the passed selection's vault
```

File: front/components/data_source_view/DataSourceViewSelector.tsx [hunk 23]
```diff
@@ -213,8 +360,32 @@ function DataSourceViewSelector({
...
+          (acc, [key, value]) =>
+            key === dataSourceView.sId ||
+            value.dataSourceView.vaultId === vaultId
+              ? { ...acc, [key]: value }
+              : acc,
+          {}
+        );
+      } else {
+        return config;
+      }
```
Comment (flvndvd) [inline front/components/data_source_view/DataSourceViewSelector.tsx:382]:
```suggestion
if (!ONLY_ONE_VAULT_PER_SELECTION) {
return config;
}

  const { vaultId, sId } = dataSourceView;
  return Object.fromEntries(
    Object.entries(config).filter(
      ([key, value]) => key === sId || value.dataSourceView.vaultId === vaultId
    )
  );
```

File: front/components/data_source_view/DataSourceViewSelector.tsx (hunk 22 above)
Comment (flvndvd) [inline front/components/data_source_view/DataSourceViewSelector.tsx:366]:
```suggestion
  // When users have multiple vaults, they can opt to select only one vault per tool.
  // This is enforced in the UI via a radio button, ensuring single selection at a time.
  // However, selecting a new item in a different vault doesn't automatically clear previous selections.
  // This function ensures that only the selections matching the current vault are retained, removing any others.
```

File: front/components/vaults/VaultLayout.tsx [hunk 24]
```diff
@@ -125,7 +121,7 @@ function VaultBreadCrumbs({
 
     const items = [
       {
-        icon: vault.kind === "global" ? CompanyIcon : LockIcon,
+        icon: getVaultIcon(vault),
```
Comment (flvndvd) [inline front/components/vaults/VaultLayout.tsx:124]: Henry introduced this help in his last PR.

**Review by flvndvd — APPROVED** [review]
LGTM, make sure to test one last time 🙏

**Comment by Duncid** [comment]
@flvndvd @Fraggle is this the flow we're talking about?

https://www.figma.com/design/QOxHwppG64GtPocpDhS6Y7?node-id=8541-57623#893198049

**Comment by Fraggle** [comment]
@Duncid yep !

**Comment by flvndvd** [comment]
We still need to do some CSS adjustments.
